src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs

```diff
@@ -138,73 +138,59 @@ private TypeSymbolWithAnnotations(TypeSymbol defaultType, INonNullTypesContext n
+        //[Obsolete] 
+        public static TypeSymbolWithAnnotations Create(TypeSymbol typeSymbol, ImmutableArray<CustomModifier> customModifiers = default)
         {
-            // PROTOTYPE(NullableReferenceTypes): Should be using fine-grained NonNullTypesContext
-            var context = isNullableIfReferenceType == null ? NonNullTypesFalseContext.Instance : NonNullTypesTrueContext.Instance;
-            return Create(typeSymbol, context, isAnnotated: isNullableIfReferenceType == true, treatUnconstrainedTypeParameterAsNullable, customModifiers.NullToEmpty());
+            return Create(NonNullTypesTrueContext.Instance, typeSymbol, customModifiers: customModifiers);
         }
 
         // PROTOTYPE(NullableReferenceTypes): Check we are not using this method on type references in
         // member signatures visible outside the assembly. Consider overriding, implementing, NoPIA embedding, etc.
-        public static TypeSymbolWithAnnotations Create(TypeSymbol typeSymbol, INonNullTypesContext nonNullTypesContext, bool isAnnotated, ImmutableArray<CustomModifier> customModifiers)
-        {
-            return Create(typeSymbol, nonNullTypesContext, isAnnotated, treatUnconstrainedTypeParameterAsNullable: true, customModifiers);
-        }
 
-        // PROTOTYPE(NullableReferenceTypes): Check we are not using this method on type references in
-        // member signatures visible outside the assembly. Consider overriding, implementing, NoPIA embedding, etc.
-        private static TypeSymbolWithAnnotations Create(TypeSymbol typeSymbol, INonNullTypesContext nonNullTypesContext, bool isAnnotated, bool treatUnconstrainedTypeParameterAsNullable, ImmutableArray<CustomModifier> customModifiers)
+        /// <summary>
+        /// Create an instance setting <see cref="IsNullable"/> directly.
+        /// </summary>
+        /// <remarks>
+        /// Can be used for scenarios where calculating <see cref="IsNullable"/> eagerly will not
+        /// result in cycles (when binding executable code for instance), or scenarios where
+        /// <see cref="IsNullable"/> is determined by state other than <see cref="IsAnnotated"/>
+        /// (in flow analysis for instance).
+        /// </remarks>
+        public static TypeSymbolWithAnnotations Create(TypeSymbol typeSymbol, bool? isNullableIfReferenceType, bool treatUnconstrainedTypeParameterAsNullable = true, ImmutableArray<CustomModifier> customModifiers = default)
         {
-            Debug.Assert(nonNullTypesContext != null);
-            Debug.Assert(!customModifiers.IsDefault);
-
             if (typeSymbol is null)
             {
                 return default;
             }
 
+            // PROTOTYPE(NullableReferenceTypes): Should be using fine-grained NonNullTypesContext
+            var context = isNullableIfReferenceType == null ? NonNullTypesFalseContext.Instance : NonNullTypesTrueContext.Instance;
```

Comment:
>context [](start = 16, length = 7)

If it were possible, I'd be tempted to use an irrelevant context instead, which is guaranteed not to be relied on. Something to consider for a follow-up.

Reply:
Added a comment.

---
In reply to: [209018603](https://github.com/dotnet/roslyn/pull/29183#discussion_r209018603) [](ancestors = 209018603)